File: backtracking/22. Generate Parentheses/Solution.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<String>();
        helper("", 0,0, n, result);
        return result;
    }

    private void helper(String generated, int open, int close, int n, List<String> result) {
        if (generated.length() == 2*n) {
            result.add(generated);
            return;
        }

        if (open < n){
            helper(generated + "(", open+1, close, n, result);
        }

        if (close < open) {
            helper(generated + ")",open, close+1, n, result);
        }
    }
}

File: backtracking/22. Generate Parentheses/Solution.go
package Solution

func generateParenthesis(n int) []string {
	// result := make([]string, )
	var result = []string{}
	helper(&result, "", 0, 0, n)
	return result
}

func helper(result *[]string, generated string, open, close, n int) {
	if len(generated) == 2*n {
		*result = append(*result, generated)
		return
	}

	if open < n {
		helper(result, generated+"(", open+1, close, n)
	}

	if close < open {
		helper(result, generated+")", open, close+1, n)
	}
}
